### PR TITLE
Add basic benchmark

### PR DESCRIPTION
**Pull request by turion**

### CODE REVIEW

**Comment by turion** [comment]
## Results

```
$ mix run benchmark.exs 
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
Number of Available Cores: 8
Available memory: 23.26 GB
Elixir 1.11.4
Erlang 23.3.1

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 0 ns
parallel: 1
inputs: large sequence, medium sequence, small sequence
Estimated total run time: 1.75 min

Benchmarking concat with input large sequence...
Benchmarking concat with input medium sequence...
Benchmarking concat with input small sequence...
Benchmarking cons with input large sequence...
Benchmarking cons with input medium sequence...
Benchmarking cons with input small sequence...
Benchmarking snoc with input large sequence...
Benchmarking snoc with input medium sequence...
Benchmarking snoc with input small sequence...
Benchmarking view_l with input large sequence...
Benchmarking view_l with input medium sequence...
Benchmarking view_l with input small sequence...
Benchmarking view_r with input large sequence...
Benchmarking view_r with input medium sequence...
Benchmarking view_r with input small sequence...

##### With input large sequence #####
Name             ips        average  deviation         median         99th %
cons          8.46 M      118.19 ns ±25740.21%           0 ns         563 ns
snoc          4.89 M      204.45 ns ±64780.06%           0 ns         655 ns
view_r        1.16 M      862.04 ns ±24324.87%         373 ns        2486 ns
view_l      0.0873 M    11457.23 ns  ±3366.98%        9623 ns       26691 ns
concat      0.0423 M    23650.48 ns  ±2346.67%       20702 ns    49741.08 ns

Comparison: 
cons          8.46 M
snoc          4.89 M - 1.73x slower +86.26 ns
view_r        1.16 M - 7.29x slower +743.85 ns
view_l      0.0873 M - 96.94x slower +11339.04 ns
concat      0.0423 M - 200.10x slower +23532.29 ns

##### With input medium sequence #####
Name             ips        average  deviation         median         99th %
cons          6.49 M      154.11 ns ±46229.80%           0 ns        1611 ns
snoc          5.95 M      168.21 ns ±10594.08%           0 ns        1751 ns
view_r        1.31 M      760.72 ns ±12704.18%         367 ns        2465 ns
view_l       0.167 M     6000.69 ns  ±1340.96%        5237 ns    10928.93 ns
concat      0.0611 M    16359.22 ns   ±928.10%       14477 ns    35965.02 ns

Comparison: 
cons          6.49 M
snoc          5.95 M - 1.09x slower +14.10 ns
view_r        1.31 M - 4.94x slower +606.61 ns
view_l       0.167 M - 38.94x slower +5846.58 ns
concat      0.0611 M - 106.15x slower +16205.11 ns

##### With input small sequence #####
Name             ips        average  deviation         median         99th %
cons          7.75 M      129.06 ns  ±6262.93%           0 ns        1710 ns
snoc          3.49 M      286.27 ns ±16267.68%           0 ns        1883 ns
view_r        1.14 M      875.78 ns ±10260.33%         475 ns        3253 ns
view_l        0.36 M     2747.03 ns  ±4271.16%        1994 ns        7109 ns
concat       0.139 M     7209.96 ns  ±2192.77%        6340 ns    14390.80 ns

Comparison: 
cons          7.75 M
snoc          3.49 M - 2.22x slower +157.21 ns
view_r        1.14 M - 6.79x slower +746.72 ns
view_l        0.36 M - 21.28x slower +2617.97 ns
concat       0.139 M - 55.86x slower +7080.90 ns
```

## Conclusion

* `concat` shows logarithmic behaviour, `cons` and `snoc` are constant.  Well done!
* `view_l` seems in fact have logarithmic runtime, see #16.
* Surprisingly, `view_r` seems to have constant runtime! What's going on here?

**Comment by turion** [comment]
I'm pretty sure the reason why `view_l` and `view_r` are so different is because of the specific shape the test cases have. `Seq.new(enumerable)` builds up the sequence with `snoc`s, so naturally it's easy to `view_r`. If you build it up in the reverse way with `cons`es, the runtime for `view_r` goes up and the runtime for `view_l` goes down.

**Comment by turion** [comment]
```
$ mix run benchmarks/interval_map.exs && mix run benchmarks/seq.exs 
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
Number of Available Cores: 8
Available memory: 23.26 GB
Elixir 1.11.4
Erlang 23.3.1

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 0 ns
parallel: 1
inputs: medium interval map, small interval map, tiny interval map
Estimated total run time: 1.05 min

Benchmarking insert with input medium interval map...
Benchmarking insert with input small interval map...
Benchmarking insert with input tiny interval map...
Benchmarking interval_match with input medium interval map...
Benchmarking interval_match with input small interval map...
Benchmarking interval_match with input tiny interval map...
Benchmarking interval_search with input medium interval map...
Benchmarking interval_search with input small interval map...
Benchmarking interval_search with input tiny interval map...

##### With input medium interval map #####
Name                      ips        average  deviation         median         99th %
interval_search     165204.96     6053.09 ns   ±173.24%        3236 ns       27912 ns
insert               17196.70    58150.70 ns    ±36.73%    53665.50 ns   146502.50 ns
interval_match          28.20 35467278.34 ns    ±28.35%    36068518 ns 57958245.80 ns

Comparison: 
interval_search     165204.96
insert               17196.70 - 9.61x slower +52097.61 ns
interval_match          28.20 - 5859.37x slower +35461225.25 ns

##### With input small interval map #####
Name                      ips        average  deviation         median         99th %
interval_search     165930.81     6026.61 ns   ±314.17%        3988 ns       22006 ns
insert               44107.90    22671.67 ns    ±54.06%       21559 ns    66440.92 ns
interval_match        3013.38   331853.43 ns    ±39.86%   324586.50 ns   813861.03 ns

Comparison: 
interval_search     165930.81
insert               44107.90 - 3.76x slower +16645.07 ns
interval_match        3013.38 - 55.06x slower +325826.82 ns

##### With input tiny interval map #####
Name                      ips        average  deviation         median         99th %
interval_search     831031.82     1203.32 ns  ±3142.45%         867 ns        3604 ns
insert              416355.17     2401.80 ns  ±1017.23%        1941 ns        7049 ns
interval_match      398987.76     2506.34 ns  ±1397.11%        1738 ns        6965 ns

Comparison: 
interval_search     831031.82
insert              416355.17 - 2.00x slower +1198.47 ns
interval_match      398987.76 - 2.08x slower +1303.02 ns
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
Number of Available Cores: 8
Available memory: 23.26 GB
Elixir 1.11.4
Erlang 23.3.1

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 0 ns
parallel: 1
inputs: medium sequence cons, medium sequence snoc, small sequence cons, small sequence snoc, tiny sequence cons, tiny sequence snoc
Estimated total run time: 3.50 min

Benchmarking concat with input medium sequence cons...
Benchmarking concat with input medium sequence snoc...
Benchmarking concat with input small sequence cons...
Benchmarking concat with input small sequence snoc...
Benchmarking concat with input tiny sequence cons...
Benchmarking concat with input tiny sequence snoc...
Benchmarking cons with input medium sequence cons...
Benchmarking cons with input medium sequence snoc...
Benchmarking cons with input small sequence cons...
Benchmarking cons with input small sequence snoc...
Benchmarking cons with input tiny sequence cons...
Benchmarking cons with input tiny sequence snoc...
Benchmarking snoc with input medium sequence cons...
Benchmarking snoc with input medium sequence snoc...
Benchmarking snoc with input small sequence cons...
Benchmarking snoc with input small sequence snoc...
Benchmarking snoc with input tiny sequence cons...
Benchmarking snoc with input tiny sequence snoc...
Benchmarking view_l with input medium sequence cons...
Benchmarking view_l with input medium sequence snoc...
Benchmarking view_l with input small sequence cons...
Benchmarking view_l with input small sequence snoc...
Benchmarking view_l with input tiny sequence cons...
Benchmarking view_l with input tiny sequence snoc...
Benchmarking view_r with input medium sequence cons...
Benchmarking view_r with input medium sequence snoc...
Benchmarking view_r with input small sequence cons...
Benchmarking view_r with input small sequence snoc...
Benchmarking view_r with input tiny sequence cons...
Benchmarking view_r with input tiny sequence snoc...

##### With input medium sequence cons #####
Name             ips        average  deviation         median         99th %
cons      1951128.60      512.52 ns  ±5605.68%         339 ns        1432 ns
snoc      1660178.26      602.34 ns  ±5718.94%         374 ns        1671 ns
view_l     914927.97     1092.98 ns  ±2338.63%         900 ns        2876 ns
view_r     148294.89     6743.32 ns   ±104.30%        5988 ns       17588 ns
concat      63248.46    15810.66 ns   ±114.72%       13360 ns    44066.75 ns

Comparison: 
cons      1951128.60
snoc      1660178.26 - 1.18x slower +89.82 ns
view_l     914927.97 - 2.13x slower +580.46 ns
view_r     148294.89 - 13.16x slower +6230.80 ns
concat      63248.46 - 30.85x slower +15298.14 ns

##### With input medium sequence snoc #####
Name             ips        average  deviation         median         99th %
cons      2086783.14      479.21 ns  ±5135.53%         320 ns        1233 ns
snoc      1606328.20      622.54 ns  ±5091.69%         388 ns        1646 ns
view_r     899138.11     1112.18 ns  ±2378.14%         871 ns        3099 ns
view_l     147674.38     6771.66 ns   ±111.09%        5920 ns       22719 ns
concat      67336.27    14850.84 ns    ±58.95%       12925 ns    39063.78 ns

Comparison: 
cons      2086783.14
snoc      1606328.20 - 1.30x slower +143.33 ns
view_r     899138.11 - 2.32x slower +632.97 ns
view_l     147674.38 - 14.13x slower +6292.45 ns
concat      67336.27 - 30.99x slower +14371.63 ns

##### With input small sequence cons #####
Name             ips        average  deviation         median         99th %
cons      1958831.53      510.51 ns  ±6368.75%         333 ns        1255 ns
snoc      1636509.78      611.06 ns  ±6487.74%         365 ns        1722 ns
view_l     841548.15     1188.29 ns  ±2667.14%         926 ns        3258 ns
view_r     327669.56     3051.86 ns   ±738.46%        2508 ns        7983 ns
concat     117024.06     8545.25 ns   ±269.65%        6913 ns       22721 ns

Comparison: 
cons      1958831.53
snoc      1636509.78 - 1.20x slower +100.55 ns
view_l     841548.15 - 2.33x slower +677.78 ns
view_r     327669.56 - 5.98x slower +2541.35 ns
concat     117024.06 - 16.74x slower +8034.74 ns

##### With input small sequence snoc #####
Name             ips        average  deviation         median         99th %
snoc      1781435.29      561.35 ns  ±6385.20%         356 ns        1484 ns
cons      1698122.05      588.89 ns  ±7069.63%         368 ns        1487 ns
view_r     912386.71     1096.03 ns  ±2653.18%         875 ns        3027 ns
view_l     266927.52     3746.34 ns   ±732.09%        2685 ns        9642 ns
concat     147750.28     6768.18 ns    ±86.22%        5908 ns       23231 ns

Comparison: 
snoc      1781435.29
cons      1698122.05 - 1.05x slower +27.54 ns
view_r     912386.71 - 1.95x slower +534.68 ns
view_l     266927.52 - 6.67x slower +3184.99 ns
concat     147750.28 - 12.06x slower +6206.83 ns

##### With input tiny sequence cons #####
Name             ips        average  deviation         median         99th %
view_l    4445423.21      224.95 ns ±15511.53%         114 ns         551 ns
view_r    4061980.46      246.19 ns ±15329.50%         122 ns         611 ns
snoc      1081677.97      924.49 ns  ±2408.60%         741 ns        2448 ns
cons      1042749.47      959.00 ns  ±2411.27%         764 ns        2413 ns
concat     951890.76     1050.54 ns  ±2056.74%         770 ns        2632 ns

Comparison: 
view_l    4445423.21
view_r    4061980.46 - 1.09x slower +21.23 ns
snoc      1081677.97 - 4.11x slower +699.54 ns
cons      1042749.47 - 4.26x slower +734.05 ns
concat     951890.76 - 4.67x slower +825.59 ns

##### With input tiny sequence snoc #####
Name             ips        average  deviation         median         99th %
view_l    4545173.12      220.01 ns ±16213.06%         109 ns         525 ns
view_r    4155073.59      240.67 ns ±15981.05%         120 ns         585 ns
snoc      1036479.36      964.80 ns  ±2232.55%         719 ns        2604 ns
cons      1031473.49      969.49 ns  ±2367.09%         751 ns        2564 ns
concat     851220.20     1174.78 ns  ±2565.54%         846 ns        2930 ns

Comparison: 
view_l    4545173.12
view_r    4155073.59 - 1.09x slower +20.66 ns
snoc      1036479.36 - 4.39x slower +744.79 ns
cons      1031473.49 - 4.41x slower +749.47 ns
concat     851220.20 - 5.34x slower +954.77 ns
```

**Comment by turion** [comment]
## Observations

* both `view_l` and `view_r` have worst case logarithmic runtime
* For interval maps, `insert` and `interval_match` seem to have the expected logarithmic runtime
* `interval_search` is surprisingly fast...

**Comment by thalesmg** [comment]
Hello! Thanks for the pull requests! I'll have a look at them soon.
Meanwhile, I noticed that Travis is not doing its part quite well and I configured Github Actions for CI.
Would you please rebase your PRs with master so those checks can run? (I hope it works with external PRs :crossed_fingers: )

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
+name: Build
```
Comment (thalesmg) [inline .github/workflows/build.yaml:1]: Strange... this shouldn't appear as a change, since it is in master. :thinking:

**Review by thalesmg — APPROVED** [review]
Thanks for the contribution! :beers:

File: .github/workflows/build.yaml (hunk 1 above)
Comment (turion) [reply .github/workflows/build.yaml]: I rebased again, better now?

**Pull request merged**